File: chains/rust/zkvm/src/block.rs
// SPDX-License-Identifier: BSD-3-Clause-Eco
// Copyright (C) 2019-2026, Lux Industries Inc. All rights reserved.

//! A block: what it says, what it is called, and what can be decided about it
//! without asking the chain anything.
//!
//! ## The id is a fold over CONTENT, and it opens with the chain
//!
//! [`compute_id`] begins with `sha256(ChainID ‖ NetworkID)`, which is NOT on
//! the wire. Two chains with different ids and an identical genesis would
//! otherwise derive the same genesis id, and one chain's blocks would then
//! chain onto the other's verbatim.
//!
//! Each transaction contributes its own content fold rather than an id read off
//! the wire — there is no such field to read — because a block whose identity
//! depended on an id a peer supplied would have as many identities as the peer
//! cared to send.
//!
//! ## The shape half of verification
//!
//! [`Block::shape`] is everything the reference decides before it reads the
//! chain, in the order it decides it: a height-0 block that names a parent, the
//! transaction cap, the clock, a nullifier repeated inside one block, and then
//! each transaction's own well-formedness and expiry. Three of those are
//! BLOCK-level and could not live in a per-transaction pass at all. What
//! remains — the proofs, the spent set, the parent, the tip, the state root —
//! needs the chain and belongs to [`crate::vm`].

use std::collections::BTreeSet;

use crate::error::{Error, Result};
use crate::hash::Fold;
use crate::host;
use crate::ids::{self, Id};
use crate::txs::Tx;
use crate::wire::{self, BlockBody};

/// How far ahead of this node's clock a block may be stamped.
pub const MAX_CLOCK_SKEW: i64 = 60;

/// The chain binding: `sha256(ChainID ‖ NetworkID)`.
///
/// It is hashed into every block and vertex id and into the public inputs every
/// shielded proof is checked against, and it is NOT on the wire — so a block or
/// a proof made for another chain does not name a block of this one and does
/// not verify here, rather than passing a check someone could forget to write.
pub fn binding(chain: &Id, network: u32) -> Id {
    let mut h = Fold::new();
    h.raw(chain).num32(network);
    h.id()
}

/// A block's identity.
pub fn compute_id(bind: &Id, b: &BlockBody) -> Id {
    let mut h = Fold::new();
    h.raw(bind).raw(&b.parent).num(b.height).signed(b.timestamp);
    for tx in &b.txs {
        h.raw(&tx.id());
    }
    h.raw(&b.state_root);
    if let Some(p) = &b.proof {
        h.raw(&p.system).raw(&p.data);
    }
    h.id()
}

/// A block of the Z-chain.
#[derive(Clone, Debug)]
pub struct Block {
    body: BlockBody,
    id: Id,
    bytes: Vec<u8>,
    status: host::Status,
}

impl Block {
    /// A block this node built, encoded and named.
    pub fn new(bind: &Id, body: BlockBody) -> Block {
        let bytes = wire::write_block(&body);
        let id = compute_id(bind, &body);
        Block {
            body,
            id,
            bytes,
            status: host::Status::Processing,
        }
    }

    /// A block off the wire.
    ///
    /// The bytes are kept AS GIVEN rather than re-encoded, which costs nothing
    /// here: the frame rule already refuses a buffer that does not account for
    /// every byte, so what came in is the one encoding of what came out.
    pub fn parse(bind: &Id, raw: &[u8]) -> Result<Block> {
        let body = wire::read_block(raw)?;
        let id = compute_id(bind, &body);
        Ok(Block {
            body,
            id,
            bytes: raw.to_vec(),
            status: host::Status::Unknown,
        })
    }

    pub fn id(&self) -> Id {
        self.id
    }

    pub fn parent(&self) -> Id {
        self.body.parent
    }

    pub fn height(&self) -> u64 {
        self.body.height
    }

    pub fn timestamp(&self) -> i64 {
        self.body.timestamp
    }

    pub fn txs(&self) -> &[Tx] {
        &self.body.txs
    }

    pub fn state_root(&self) -> &[u8] {
        &self.body.state_root
    }

    pub fn body(&self) -> &BlockBody {
        &self.body
    }

    pub fn bytes(&self) -> &[u8] {
        &self.bytes
    }

    pub fn status(&self) -> host::Status {
        self.status
    }

    pub fn set_status(&mut self, s: host::Status) {
        self.status = s;
    }

    /// What a Z block IS, is what it CARRIES.
    ///
    /// The chain has one block type, so naming the block names nothing — the
    /// word would be the same on every row. A block of one kind repeated is
    /// named once with its count; a mixed block spells every one out, because
    /// which kinds a block mixes is the thing worth comparing.
    pub fn carries(&self) -> String {
        let Some(first) = self.body.txs.first() else {
            return "Empty".into();
        };
        let first = first.kind.to_string();
        let mut uniform = true;
        let mut joined = first.clone();
        for tx in &self.body.txs[1..] {
            let n = tx.kind.to_string();
            if n != first {
                uniform = false;
            }
            joined.push('+');
            joined.push_str(&n);
        }
        if !uniform {
            return joined;
        }
        if self.body.txs.len() == 1 {
            return first;
        }
        format!("{first}x{}", self.body.txs.len())
    }

    /// Everything this block can be refused for without reading the chain, in
    /// the order the reference refuses it.
    ///
    /// `cap` is what a block may carry and `now` is this node's clock. A block
    /// off the wire is held to the bound a block this node builds is held to,
    /// so a proposer cannot produce one its own peers refuse.
    pub fn shape(&self, cap: u32, now: i64) -> Result<()> {
        if self.body.height == 0 && !ids::is_empty(&self.body.parent) {
            // Height 0 is genesis, and genesis has no parent. A height-0 block
            // that names one is two claims about which block it is.
            return Err(Error::InvalidBlock);
        }

        if self.body.txs.len() as u64 > u64::from(cap) {
            return Err(Error::TxCap {
                txs: self.body.txs.len(),
                cap,
            });
        }

        if self.body.timestamp > now.saturating_add(MAX_CLOCK_SKEW) {
            return Err(Error::FutureBlock);
        }

        // Every nullifier in the block must be distinct. Per-transaction
        // verification only sees nullifiers already spent in ACCEPTED state, so
        // without this two transactions in one block — or one transaction
        // listing a nullifier twice — spend the same shielded note and inflate
        // supply. Checked before the proofs because it is the cheaper gate.
        let mut here: BTreeSet<&[u8]> = BTreeSet::new();
        for tx in &self.body.txs {
            for n in &tx.nullifiers {
                if !here.insert(n.as_slice()) {
                    return Err(Error::DuplicateNullifier);
                }
            }
        }
        Ok(())
    }
}

impl host::Block for Block {
    fn id(&self) -> host::Id {
        self.id
    }

    fn parent(&self) -> host::Id {
        self.body.parent
    }

    fn height(&self) -> u64 {
        self.body.height
    }

    fn timestamp(&self) -> u64 {
        self.body.timestamp.max(0) as u64
    }

    fn bytes(&self) -> Vec<u8> {
        self.bytes.clone()
    }

    /// The state this block leaves behind: the chain's own committed root,
    /// which is a field of the block rather than something derived here. A
    /// certificate over a root nobody computed certifies nothing.
    fn state_root(&self) -> host::Id {
        ids::from_slice(&self.body.state_root)
    }

    /// What this block CARRIES: its transactions, folded in order. Separate
    /// from the state root, because a certificate that named a state but not
    /// the payload that produced it would be one two different blocks could
    /// satisfy.
    fn payload_root(&self) -> host::Id {
        let mut folded = Vec::with_capacity(self.body.txs.len() * ids::ID_LEN);
        for tx in &self.body.txs {
            folded.extend_from_slice(&tx.id());
        }
        crate::hash::sha256(&folded)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::txs::{Kind, Proof, Shielded, Tx};

    fn bind() -> Id {
        binding(&ids::repeated(40), 1)
    }

    fn transfer(nullifier: u8) -> Tx {
        Tx {
            kind: Kind::TRANSFER,
            version: 1,
            nullifiers: vec![vec![nullifier; 32]],
            outputs: vec![Shielded {
                commitment: vec![0x50; 32],
                ..Shielded::default()
            }],
            proof: Some(Proof {
                system: b"stark".to_vec(),
                data: vec![0x60; 192],
                public: vec![vec![0x61; 32]],
            }),
            fee: 1,
            expiry: 1000,
            ..Tx::default()
        }
    }

    fn body(height: u64, parent: Id, txs: Vec<Tx>) -> BlockBody {
        BlockBody {
            parent,
            height,
            timestamp: 1000,
            txs,
            state_root: vec![0xAB; 32],
            proof: None,
        }
    }

    /// The corpus's contract, computed here and asserted against the number the
    /// Go generator recorded. Both numbers are off the wire in neither case:
    /// they are the chain the evaluator was built for.
    #[test]
    fn the_binding_is_the_chain_and_the_network_and_nothing_else() {
        assert_eq!(
            ids::hex(&bind()),
            "c360d9f00bd7a0f2624fc5897800c2e24811f553b0df0e88d682d99e112acd78"
        );
    }

    /// The genesis of the chain the Z vectors are built for: height 0, no
    /// parent, the timestamp the genesis file names, no transactions and no
    /// state root. Every vector names this block as its parent, so this id is
    /// the corpus's contract and not an implementation detail.
    #[test]
    fn the_corpus_genesis_takes_the_id_the_reference_recorded() {
        let genesis = BlockBody {
            parent: ids::EMPTY,
            height: 0,
            timestamp: 1000,
            txs: Vec::new(),
            state_root: Vec::new(),
            proof: None,
        };
        assert_eq!(
            ids::hex(&compute_id(&bind(), &genesis)),
            "9c227870184f9269e1c720c2df8256dea664debf56f8eec76acab03c4bea755d"
        );
    }

    /// The same block on another chain is another block. This is the whole
    /// reason the binding is hashed in.
    #[test]
    fn the_same_bytes_name_a_different_block_on_a_different_chain() {
        let b = body(1, ids::repeated(9), vec![]);
        assert_ne!(
            compute_id(&bind(), &b),
            compute_id(&binding(&ids::repeated(41), 1), &b)
        );
        assert_ne!(
            compute_id(&bind(), &b),
            compute_id(&binding(&ids::repeated(40), 2), &b)
        );
    }

    #[test]
    fn every_field_of_a_block_is_in_its_identity() {
        let base = body(1, ids::repeated(9), vec![transfer(0x10)]);
        let id = compute_id(&bind(), &base);

        let mut moved = Vec::new();
        for f in [
            (|b: &mut BlockBody| b.parent = ids::repeated(8)) as fn(&mut BlockBody),
            |b: &mut BlockBody| b.height = 2,
            |b: &mut BlockBody| b.timestamp = 1001,
            |b: &mut BlockBody| b.state_root = vec![0xAC; 32],
            |b: &mut BlockBody| b.txs[0].nullifiers[0][0] ^= 1,
            |b: &mut BlockBody| {
                b.proof = Some(Proof {
                    system: b"stark".to_vec(),
                    data: vec![1],
                    public: vec![],
                })
            },
        ] {
            let mut b = base.clone();
            f(&mut b);
            moved.push(compute_id(&bind(), &b));
        }
        for m in moved {
            assert_ne!(id, m);
        }
    }

    #[test]
    fn a_parsed_block_keeps_the_bytes_it_was_given() {
        let raw = wire::write_block(&body(1, ids::repeated(9), vec![transfer(0x10)]));
        let b = Block::parse(&bind(), &raw).unwrap();
        assert_eq!(b.bytes(), raw.as_slice());
        assert_eq!(b.id(), compute_id(&bind(), b.body()));
    }

    #[test]
    fn a_built_block_and_the_same_block_parsed_back_are_one_block() {
        let built = Block::new(&bind(), body(1, ids::repeated(9), vec![transfer(0x10)]));
        let parsed = Block::parse(&bind(), built.bytes()).unwrap();
        assert_eq!(built.id(), parsed.id());
        assert_eq!(built.bytes(), parsed.bytes());
    }

    #[test]
    fn what_a_block_carries_is_what_it_is_called() {
        let b = |txs: Vec<Tx>| Block::new(&bind(), body(1, ids::repeated(9), txs)).carries();
        assert_eq!(b(vec![]), "Empty");
        assert_eq!(b(vec![transfer(1)]), "Transfer");
        assert_eq!(b(vec![transfer(1), transfer(2)]), "Transferx2");

        let mut shield = transfer(3);
        shield.kind = Kind::SHIELD;
        assert_eq!(b(vec![transfer(1), shield.clone()]), "Transfer+Shield");

        let mut unknown = transfer(4);
        unknown.kind = Kind(11);
        assert_eq!(b(vec![unknown]), "unknown");
    }

    #[test]
    fn a_height_zero_block_that_names_a_parent_is_two_claims() {
        let b = Block::new(&bind(), body(0, ids::repeated(9), vec![]));
        assert_eq!(b.shape(100, 1000), Err(Error::InvalidBlock));

        let genesis = Block::new(&bind(), body(0, ids::EMPTY, vec![]));
        assert_eq!(genesis.shape(100, 1000), Ok(()));
    }

    #[test]
    fn a_block_over_the_cap_is_refused_by_the_bound_a_proposer_builds_to() {
        let txs: Vec<Tx> = (0..4u8).map(transfer).collect();
        let b = Block::new(&bind(), body(1, ids::repeated(9), txs));
        assert_eq!(b.shape(4, 1000), Ok(()));
        assert_eq!(b.shape(3, 1000), Err(Error::TxCap { txs: 4, cap: 3 }));
    }

    #[test]
    fn a_block_stamped_past_the_skew_allowance_is_refused() {
        let mut body = body(1, ids::repeated(9), vec![]);
        body.timestamp = 1 << 40;
        let b = Block::new(&bind(), body);
        assert_eq!(b.shape(100, 1000), Err(Error::FutureBlock));
        // Inside the allowance it is not the clock that refuses it.
        assert_eq!(b.shape(100, (1 << 40) - MAX_CLOCK_SKEW), Ok(()));
    }

    /// One note spent twice, across two transactions that are each well formed
    /// on their own. Nothing else in the chain sees this: per-transaction
    /// verification reads only what is already spent in accepted state.
    #[test]
    fn one_note_spent_twice_inside_one_block_is_refused() {
        let mut twin = transfer(0x10);
        twin.fee = 2;
        let b = Block::new(&bind(), body(1, ids::repeated(9), vec![transfer(0x10), twin]));
        assert_eq!(b.shape(100, 1000), Err(Error::DuplicateNullifier));
    }

    #[test]
    fn one_transaction_listing_a_nullifier_twice_is_refused_the_same_way() {
        let mut tx = transfer(0x10);
        tx.nullifiers.push(vec![0x10; 32]);
        let b = Block::new(&bind(), body(1, ids::repeated(9), vec![tx]));
        assert_eq!(b.shape(100, 1000), Err(Error::DuplicateNullifier));
    }

    /// The order the reference refuses in. A block that breaks two rules is
    /// refused for the first one, and a port that reordered them would report a
    /// different reason for the same block.
    #[test]
    fn the_shape_rules_are_checked_in_the_reference_order() {
        let mut twin = transfer(0x10);
        twin.fee = 2;
        let mut b = body(0, ids::repeated(9), vec![transfer(0x10), twin]);
        b.timestamp = 1 << 40;
        let b = Block::new(&bind(), b);
        // Genesis-with-a-parent, over the cap, past the clock, and a repeated
        // nullifier — all four at once.
        assert_eq!(b.shape(1, 1000), Err(Error::InvalidBlock));

        let mut ok_genesis = b.body().clone();
        ok_genesis.height = 1;
        let b = Block::new(&bind(), ok_genesis);
        assert_eq!(b.shape(1, 1000), Err(Error::TxCap { txs: 2, cap: 1 }));
        assert_eq!(b.shape(100, 1000), Err(Error::FutureBlock));
        assert_eq!(b.shape(100, 1 << 40), Err(Error::DuplicateNullifier));
    }

    #[test]
    fn the_host_reads_the_block_the_chain_wrote() {
        let b = Block::new(&bind(), body(1, ids::repeated(9), vec![transfer(0x10)]));
        assert_eq!(host::Block::id(&b), b.id);
        assert_eq!(host::Block::height(&b), 1);
        assert_eq!(host::Block::timestamp(&b), 1000);
        assert_eq!(host::Block::state_root(&b), ids::repeated(0xAB));
        assert_eq!(
            host::Block::payload_root(&b),
            crate::hash::sha256(&transfer(0x10).id())
        );
        assert_eq!(b.bytes(), <Block as host::Block>::bytes(&b).as_slice());
    }
}

File: chains/cpp/zkvm/src/block.cpp
// Copyright (C) 2026, Lux Industries Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Eco

#include "lux/zkvm/block.hpp"

#include "lux/zkvm/vm.hpp"
#include "lux/zkvm/zap.hpp"

#include <set>

namespace lux::zkvm {
namespace {

// Block: ParentID 32B@0, Height u64@32, Timestamp i64@40, TxLens list@48,
//        TxBlob bytes@56, StateRoot bytes@64
constexpr int kParent = 0, kHeight = 32, kTime = 40, kTxLens = 48, kTxBlob = 56, kStateRoot = 64;

// Genesis: Timestamp i64@0, TxLens list@8, TxBlob bytes@16
constexpr int kGenTime = 0, kGenTxLens = 8, kGenTxBlob = 16;
constexpr int kGenSize = 24;

}  // namespace

Id Block::compute_id() const {
    Hasher h;
    if (vm_) h.write(view(vm_->bind()));
    h.write(view(parent_id));
    h.num(block_height);
    h.num(block_timestamp);
    for (const auto& tx : txs) {
        const Id tx_id = tx.compute_id();
        h.write(view(tx_id));
    }
    h.write(view(state_root));
    return h.sum();
}

lux::node::Id Block::id() const {
    if (id_ == kEmptyId) id_ = compute_id();
    return id_;
}

std::span<const std::uint8_t> Block::bytes() const {
    if (bytes_.empty()) bytes_ = marshal();
    return view(bytes_);
}

lux::node::Id Block::root() const {
    Id r{};
    if (state_root.size() == 32) std::copy(state_root.begin(), state_root.end(), r.begin());
    return r;
}

Bytes Block::marshal() const {
    std::vector<std::uint32_t> tx_lens;
    Bytes tx_blob;
    wire::pack_objs(txs, [](const Transaction& t) { return t.marshal(); }, tx_lens, tx_blob);

    zap::Builder b(zap::kHeaderSize + kBlkSize + int(tx_blob.size()) + int(state_root.size()) +
                   4 * int(tx_lens.size()) + 256);
    const int tx_off = wire::write_u32_list(b, tx_lens);
    auto ob = b.start_object(kBlkSize);
    ob.set_bytes_fixed(kParent, view(parent_id));
    ob.set_u64(kHeight, block_height);
    ob.set_u64(kTime, static_cast<std::uint64_t>(block_timestamp));
    ob.set_list(kTxLens, tx_off, int(tx_lens.size()));
    ob.set_bytes(kTxBlob, view(tx_blob));
    ob.set_bytes(kStateRoot, view(state_root));
    ob.finish_as_root();
    return b.finish();
}

wire::Result<void> parse_block_bytes(ByteView data, Block& blk) {
    zap::Message msg;
    auto o = wire::parse_frame(data, &msg);
    if (!o) return std::unexpected(o.error());

    blk.parent_id = wire::read_id(*o, kParent);
    blk.block_height = o->u64(kHeight);
    blk.block_timestamp = static_cast<std::int64_t>(o->u64(kTime));
    blk.state_root = wire::cp(o->bytes(kStateRoot));

    auto txs = wire::unpack_objs<Transaction>(wire::read_u32_list(*o, kTxLens),
                                              o->bytes(kTxBlob), parse_transaction);
    if (!txs) return std::unexpected(txs.error());
    blk.txs = std::move(*txs);
    return {};
}

wire::Result<void> Block::syntactic_verify() const {
    if (vm_ == nullptr) return std::unexpected("zkvm: block is not bound to a chain");

    // Height 0 is genesis, and genesis has no parent. A height-0 block that
    // names one makes two claims about which block it is.
    if (block_height == 0 && parent_id != kEmptyId) return std::unexpected(kErrInvalidBlock);

    // A block off the wire is held to the bound a block this node builds is held
    // to, so a proposer cannot produce one its own peers refuse. The bound is
    // configuration, fixed at genesis — no block moves it.
    if (txs.size() > vm_->z_config().max_tx_per_block)
        return std::unexpected(std::string(kErrInvalidBlock) + ": " +
                               std::to_string(txs.size()) + " transactions over the " +
                               std::to_string(vm_->z_config().max_tx_per_block) + " cap");

    // The node's own clock, not the ledger. Held against the parent's timestamp
    // too, but that is a different rule and it lives below with the parent.
    if (block_timestamp > vm_->now() + kMaxClockSkew) return std::unexpected(kErrFutureBlock);

    // Block-level shape: every nullifier in the block must be distinct.
    // verify_transaction only sees nullifiers already spent in ACCEPTED state,
    // so without this two transactions in one block — or one transaction listing
    // a nullifier twice — spend the same shielded note and inflate supply. The
    // block's transactions against each other, so the spent set is not consulted.
    ByteSet spent_here;
    for (const auto& tx : txs) {
        for (const auto& n : tx.nullifiers) {
            if (!spent_here.insert(n).second) return std::unexpected(kErrDuplicateNullifier);
        }
    }

    // Each transaction's own shape, and its expiry against the height THIS block
    // claims — which is on the wire in front of us, so no parent is needed to
    // know a transaction has outlived its window.
    for (const auto& tx : txs) {
        if (auto r = vm_->syntactic_verify(tx, block_height); !r) return r;
    }
    return {};
}

wire::Result<void> Block::check() {
    // Everything decidable from the block in hand goes first. A peer's block
    // that cannot be true of any chain is refused before this node reads a
    // single key, and the mempool asks the same question of a transaction
    // before one is ever assembled.
    if (auto r = syntactic_verify(); !r) return r;

    // Then the ledger. admit re-asks the shape half above, which has already
    // passed: a handful of size comparisons next to a STARK, and the price of
    // check asking assembly's predicate VERBATIM rather than a copy of its two
    // halves that a later edit could let drift.
    for (const auto& tx : txs) {
        if (auto r = vm_->admit(tx, block_height); !r) return r;
    }

    if (block_height > 0) {
        auto parent = vm_->block(parent_id);
        if (!parent) return std::unexpected(parent.error());

        // The parent must be one this chain can still build on: the accepted
        // tip, or a block verified above it and not yet decided. Height alone is
        // not that check — a block whose parent is an OLD accepted block
        // satisfies height == parent+1 perfectly well, and accepting it rewinds
        // the tip and leaves the height index naming an orphan as the block at
        // that height to every peer that bootstraps from it.
        const Id tip = vm_->chain().tip();
        const std::uint64_t tip_height = vm_->chain().height();
        if ((*parent)->id() != tip && (*parent)->block_height <= tip_height)
            return std::unexpected(std::string(kErrNotOnTip) + ": parent " +
                                   hex((*parent)->id()) + " at height " +
                                   std::to_string((*parent)->block_height) +
                                   " is beneath the tip at " + std::to_string(tip_height));

        if (block_height != (*parent)->block_height + 1)
            return std::unexpected(kErrInvalidHeight);
        if (block_timestamp < (*parent)->block_timestamp)
            return std::unexpected(kErrInvalidTimestamp);
    }

    const Id want = vm_->compute_state_root(txs);
    if (state_root.size() != want.size() ||
        !std::equal(state_root.begin(), state_root.end(), want.begin()))
        return std::unexpected(kErrInvalidStateRoot);

    // A block that verifies is one the engine may build on, so it has to be
    // findable by id — including one parsed from a peer rather than built here.
    // Tracking only self-built blocks leaves a follower able to verify the first
    // block of a run and unable to verify the second.
    vm_->chain().track(std::static_pointer_cast<Decision>(shared_from_this()));
    return {};
}

bool Block::verify() {
    auto r = check();
    if (!r) {
        error_ = r.error();
        return false;
    }
    error_.clear();
    return true;
}

// commit applies the block. Everything is staged and committed in one batch with
// the block and the tip, so a spend that cannot be recorded takes the whole block
// with it.
//
// This used to mark the block accepted and move the tip before writing anything,
// then issue one write per nullifier and per output, each returning early. A
// failure partway left some notes spent and some outputs created, under a tip the
// chain had already advanced — a shielded pool half applied, with no way back and
// no way to apply the block again.
wire::Result<void> Block::commit() {
    if (vm_ == nullptr) return std::unexpected("zkvm: block is not bound to a chain");
    return vm_->chain().accept(std::static_pointer_cast<Decision>(shared_from_this()));
}

void Block::accept() {
    // Decided once. A block already rejected can never be accepted, and one
    // already accepted is durable — re-running the commit would be asked to
    // spend notes this block has already spent.
    if (status_ != Status::Processing) return;
    auto r = commit();
    if (!r) error_ = r.error();
}

wire::Result<void> Block::write(store::Store& view_store) {
    (void)view_store;  // the three stores were built over this same view
    for (const auto& tx : txs) {
        for (const auto& n : tx.nullifiers) {
            if (auto r = vm_->nullifiers().mark_spent(view(n), block_height); !r) return r;
        }
        for (std::size_t i = 0; i < tx.outputs.size(); ++i) {
            Utxo u;
            u.tx_id = tx.id;
            u.output_index = std::uint32_t(i);
            u.commitment = tx.outputs[i].commitment;
            u.ciphertext = tx.outputs[i].encrypted_note;
            u.ephemeral_pk = tx.outputs[i].ephemeral_pubkey;
            u.height = block_height;
            if (auto r = vm_->utxos().add(u); !r) return r;
        }
    }
    Id next{};
    if (state_root.size() == 32) std::copy(state_root.begin(), state_root.end(), next.begin());
    return vm_->state_root().finalize(next);
}

// publish marks the block accepted and releases the transactions it carried. It
// runs after the commit, so a transaction is only dropped from the pool once the
// block that spends it is durable.
void Block::publish() {
    status_ = Status::Accepted;
    for (const auto& tx : txs) vm_->mempool().remove(tx.id);

    // The chain has passed this height, so anything expiring at or below it can
    // never enter a block. Nothing else drops those, and a pool full of them
    // refuses every honest arrival paying the same floor.
    vm_->mempool().prune_expired(block_height);
}

// reject hands back what the block was carrying. The transactions were never
// refused — they lost a race — so they go back into the pool, exactly as the
// reference's Reject does.
//
// It is inert on a block that is already decided. Rejecting an ACCEPTED block
// would return transactions whose notes this chain has already spent, and the
// next block this node assembled would carry a double spend that every peer
// refuses.
void Block::reject() {
    if (status_ != Status::Processing) return;
    status_ = Status::Rejected;
    vm_->chain().drop(id());
    for (const auto& tx : txs) (void)vm_->mempool().add(tx);
}

Bytes Genesis::marshal() const {
    std::vector<std::uint32_t> tx_lens;
    Bytes tx_blob;
    wire::pack_objs(initial_txs, [](const Transaction& t) { return t.marshal(); }, tx_lens,
                    tx_blob);

    zap::Builder b(zap::kHeaderSize + kGenSize + int(tx_blob.size()) + 4 * int(tx_lens.size()) +
                   128);
    const int tx_off = wire::write_u32_list(b, tx_lens);
    auto ob = b.start_object(kGenSize);
    ob.set_u64(kGenTime, static_cast<std::uint64_t>(timestamp));
    ob.set_list(kGenTxLens, tx_off, int(tx_lens.size()));
    ob.set_bytes(kGenTxBlob, view(tx_blob));
    ob.finish_as_root();
    return b.finish();
}

wire::Result<Genesis> parse_genesis(ByteView data) {
    // An empty configuration is a chain starting at timestamp 0 with nothing
    // allocated — which is a chain, and is what the reference makes of empty
    // genesis bytes.
    if (data.empty()) return Genesis{};

    zap::Message msg;
    auto o = wire::parse_frame(data, &msg);
    if (!o) return std::unexpected(o.error());

    Genesis g;
    g.timestamp = static_cast<std::int64_t>(o->u64(kGenTime));
    auto txs = wire::unpack_objs<Transaction>(wire::read_u32_list(*o, kGenTxLens),
                                              o->bytes(kGenTxBlob), parse_transaction);
    if (!txs) return std::unexpected(txs.error());
    g.initial_txs = std::move(*txs);
    return g;
}

BlockSummary summarize(const Block& b) {
    BlockSummary s;
    s.id = b.id();
    s.height = b.block_height;
    s.timestamp = b.block_timestamp;
    s.tx_count = b.txs.size();
    s.state_root = b.state_root;
    return s;
}

}  // namespace lux::zkvm
